### PR TITLE
Fixed : #29 Multiple playlists added

### LINKED ISSUE #29: Add multiple playlists
I would like to work on this issue and provide a PR for the same.

### PR DESCRIPTION
**Pull request by aniketgattani**
Fixed #29 Uptill now I have implemented done the following:
1. Changed the cached variables.
2. Added the dropdown and playlist selection handlers.

### CODE REVIEW

**Comment by aniketgattani** [comment]
@udiboy1209 can u please review my PR

File: js/player.js [hunk 1]
```diff
@@ -47,8 +47,9 @@ function($, playlist, ytIframe){
     }
 
     var playSong = function(id){
+        playing=parseInt(localStorage.getItem('playing'));
         if(playing>=0 && playing<playlist.size() && id==playlist.get(playing).id.videoId)
-            return;
+            {console.log('svdsdv'); return;}
```
Comment (udiboy1209) [inline js/player.js]: Leaving around vague debug statements, not cool!

File: js/player.js [hunk 2]
```diff
@@ -163,10 +164,8 @@ function($, playlist, ytIframe){
         savePlaying: savePlaying,
         setPlaying: setPlaying,
         getCurrent: function() {
+            playing=parseInt(localStorage.getItem('playing'));
```
Comment (udiboy1209) [inline js/player.js]: Why is this needed? isn't the `playing` variable reliable?

File: js/player.js [hunk 3]
```diff
@@ -163,10 +164,8 @@ function($, playlist, ytIframe){
...
             return playlist.get(playing);
-        },
-        getPlaying: function(){
-            return playing;
```
Comment (udiboy1209) [inline js/player.js]: Don't undo the change introduced in #32. Because it is merged, you should rebase to master.

File: js/playlist.js [hunk 4]
```diff
@@ -1,18 +1,20 @@
...
     var playlistChangeListeners = []
-
+    var curr_id;
     var setup = function() {
+        console.log($.query);
         // Load playlist from cache/url
         if($.query.playlist!=undefined){
             ytData.api.videos.list({
-                id:$.query.playlist,
+                id:$.query.playlist[curr_id],
```
Comment (udiboy1209) [inline js/playlist.js]: Firstly, i don't see where you have changed the share link generation code to reflect as list of playlists. So I don't think this will work. I know my project does not have tests, but you should ensure as a rule of thumb that you once manually test every line of change that you make. Remember that this is a URL param, so this will be a list only if it is set in a specific way.

Secondly, the user generally wouldn't want to share each one of his playlists, only the current one. So the change here should rather be that `$.query.playlist` is added to the user's list of playlists as a new temp playlist.

File: js/playlist.js [hunk 5]
```diff
@@ -21,9 +23,12 @@ function(ytData,ytIframe,template,$) {
         } else {
             if(JSON.parse(localStorage.getItem('new_playlist_saved'))) {
                 playlist = JSON.parse(localStorage.getItem('playlist'));
-                var rptst = localStorage.getItem('repeat');
-
-                total_duration = JSON.parse(localStorage.getItem('total_duration'));
+                if (curr_id==undefined) curr_id=playlist.length-1;
+                console.log(curr_id);
+                var rptst = playlist[curr_id]['repeat'];
```
Comment (udiboy1209) [inline js/playlist.js]: Here you access `playlist[curr_id]` as a dict, whereas later in the loop, you are accessing it as a list. Have you tested this code?

File: js/playlist.js [hunk 6]
```diff
@@ -198,7 +207,42 @@ function(ytData,ytIframe,template,$) {
...
+        if(ytIframe.player!=undefined) ytIframe.player.stopVideo();
+        curr_id=parseInt(event.target.id);
+        localStorage.setItem('curr_id',curr_id);
+        $('#playlist li').remove();
+        localStorage.setItem('playing',-1);
+        setup();
+    }
+
     setup();
+    addPlaylistDropdown();
```
Comment (udiboy1209) [inline js/playlist.js]: Move this to `setup()`

**Comment by aniketgattani** [comment]
@udiboy1209 thanks a lot for pointing out these. I had made these changes mostly to know ur review as soon as possible so that I may proceed accordingly. I will make the changes asap. Also even i want to make the playlist name customizable to the user. But I will make that commit a little later. I wanted to see that I am going in the right dirn or not and also I did check the code. It was running well. There were some parts which I was not sure of. But I havent received much of the help from the gitter chats as well. Thats why I was a bit struck :\ and cudnt quite test all the changes :\ .

**Comment by aniketgattani** [comment]
@udiboy1209 I have made the changes please see them now. I think everything should be fine now. I will also make the playlist names customizable by giving a different PR.

File: js/playlist.js [hunk 7]
```diff
@@ -1,9 +1,11 @@
 define(['ytDataAPI', 'ytIframeAPI', 'template','jquery'],
 function(ytData,ytIframe,template,$) {
     var playlist=[];
+    var durations=[];
+    var repeat = [];
```
Comment (udiboy1209) [inline js/playlist.js]: I don't feel that the repeat option should be specific to every playlist. Because it is more of a choice on how the user wants to listen to the current playlist in general. I can't think of a case when a user may want to loop for a certain playlist, and not loop or shuffle for a different one.

File: js/playlist.js [hunk 8]
```diff
@@ -21,36 +23,41 @@ function(ytData,ytIframe,template,$) {
         } else {
             if(JSON.parse(localStorage.getItem('new_playlist_saved'))) {
                 playlist = JSON.parse(localStorage.getItem('playlist'));
-                var rptst = localStorage.getItem('repeat');
-
-                total_duration = JSON.parse(localStorage.getItem('total_duration'));
+                if (curr_id==undefined) curr_id=playlist.length-1;
```
Comment (udiboy1209) [inline js/playlist.js]: what if `playlist.length` is 0. That will lead to a crash further on.

File: js/playlist.js [hunk 9]
```diff
@@ -21,36 +23,41 @@ function(ytData,ytIframe,template,$) {
...
-            $(vidrow).addClass("item-video waves-effect");
-            $(vidrow).on("click", function(id){return function(event){
-                require('player').playSong(id);
-            }}(id));
-            $(vidrow).find("#action").on("click",function(id){return function(event){
-                removeFromPlaylist(event,id);
-            }}(id));
-            $(vidrow).find("#action i").html("clear");
-            vidrow.appendTo("#playlist");
+        if (playlist.length!=0){
```
Comment (udiboy1209) [inline js/playlist.js]: To avoid these checks and unnecessary crashes like the one I pointed to above, ensure that there is atleast one playlist at all times in the playlist array, even if its an empty one. i.e. `playlist = [[]]`. Never let the `playlist.length` become 0. It seems pointless to care for a case when there are "no playlists", instead that case can be considered as one empty playlist.

File: js/playlist.js [hunk 10]
```diff
@@ -94,30 +101,36 @@ function(ytData,ytIframe,template,$) {
...
+        durations[curr_id]=total_duration;
+        localStorage.setItem('duration', JSON.stringify(durations));
     }
 
     var savePlaylist = function() {
         localStorage.setItem('playlist', JSON.stringify(playlist));
         localStorage.setItem('new_playlist_saved', JSON.stringify(true));
     }
 
+    var saveRepeat = function(mode){
```
Comment (udiboy1209) [inline js/playlist.js]: If you change to having a common repeat option for every playlist, let it stay in `player.js`, because repeat option is a player specific property not a playlist specific property.

**Comment by udiboy1209** [comment]
Seems like this PR has been abandoned, so I'm closing this.
